Mathlib/FieldTheory/Perfect.lean

```diff
@@ -218,3 +218,17 @@ instance toPerfectRing (p : ℕ) [hp : Fact p.Prime] [CharP K p] : PerfectRing K
   exact minpoly.degree_pos ha
 
 end PerfectField
+
+/-- If `L / K` is an algebraic extension, `K` is a perfect field, then `L / K` is separable. -/
+theorem Algebra.IsAlgebraic.isSeparable_of_perfectField {K L : Type*} [Field K] [Field L]
+    [Algebra K L] [PerfectField K] (halg : Algebra.IsAlgebraic K L) : IsSeparable K L :=
+  ⟨fun x ↦ PerfectField.separable_of_irreducible <| minpoly.irreducible (halg x).isIntegral⟩
+
+/-- If `L / K` is an algebraic extension, `K` is a perfect field, then so is `L`. -/
+theorem Algebra.IsAlgebraic.perfectField {K L : Type*} [Field K] [Field L] [Algebra K L]
+    [PerfectField K] (halg : Algebra.IsAlgebraic K L) : PerfectField L := ⟨fun {f} hf ↦
+  haveI := Fact.mk hf
+  have h := hf.ne_zero
+  have h2 := isIntegral_trans halg.isIntegral _ (AdjoinRoot.isIntegral_root h)
+  have h3 := (AdjoinRoot.minpoly_root h) ▸ minpoly.dvd_map_of_isScalarTower K L (AdjoinRoot.root f)
+  (PerfectField.separable_of_irreducible (minpoly.irreducible h2)).map |>.of_dvd h3 |>.of_mul_left⟩
```

Comment:
Can you extract the lemma that says if L/K is integral with L a field and K a CommRing, then any irreducible polynomial in L[X] divides some monic irreducible polynomial in K[X]?

I think we can also prove it for an integrally closed domain L if we assume `Monic` because we have [minpoly.isIntegrallyClosed_dvd](https://leanprover-community.github.io/mathlib4_docs/Mathlib/FieldTheory/Minpoly/IsIntegrallyClosed.html#minpoly.isIntegrallyClosed_dvd) and [IsIntegrallyClosed.minpoly.unique](https://leanprover-community.github.io/mathlib4_docs/Mathlib/FieldTheory/Minpoly/IsIntegrallyClosed.html#IsIntegrallyClosed.minpoly.unique). For L/K free maybe we can use Algebra.norm, but I'm not sure whether every element divides its norm in a general algebra; even though we can factor into irreducibles in a Noetherian domain, we probably need to start with a prime (not just irreducible) polynomial in L[X] in order for it to divide some factor ... there could be a version that assumes IsAlgebraic instead of IsIntegral too. But not for this PR ...

Reply:
Sure, let me check later.

Reply:
> Can you extract the lemma that says if L/K is integral with L a field and K a CommRing, then any irreducible polynomial in L[X] divides some monic irreducible polynomial in K[X]?

Need `IsDomain K`. Can it be deduced from existing conditions?

Reply:
I see. There's no obvious way to remove IsDomain as far as I can see. If K is not a domain, it's possible that the minpoly has higher degree factors that become units in L[X].
Let's move this lemma to the file where AdjoinRoot is defined and call it a day!

Reply:
> > Can you extract the lemma that says if L/K is integral with L a field and K a CommRing, then any irreducible polynomial in L[X] divides some monic irreducible polynomial in K[X]?
> 
> Need `IsDomain K` (used in `minpoly.irreducible`).

This result is incorrect, at least when `K` is not connected (i.e. can be written as a product of two rings), see https://math.stackexchange.com/a/4849403/235999. On the other hand, it's true if `K` is irreducible (i.e. only one minimal prime ideal), see https://math.stackexchange.com/a/4843432/235999.